include/revng/Model/Types.h

```diff
@@ -0,0 +1,903 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <algorithm>
+#include <compare>
+#include <optional>
+
+#include "llvm/Support/RandomNumberGenerator.h"
+
+#include "revng/ADT/KeyedObjectTraits.h"
+#include "revng/ADT/RecursiveCoroutine.h"
+#include "revng/ADT/SortedVector.h"
+#include "revng/ADT/UpcastablePointer.h"
+#include "revng/ADT/UpcastablePointer/YAMLTraits.h"
+#include "revng/Model/TupleTree.h"
+#include "revng/Support/Assert.h"
+#include "revng/Support/Debug.h"
+
+namespace flatc {
+
+namespace TypeKind {
+
+/// \brief Enum for identifying different kind of flatc types.
+enum Values {
+  Invalid,
+  Primitive,
+  Enum,
+  Typedef,
+  Struct,
+  Union,
+  FunctionPtr,
+};
+// TODO: fromString toString
+} // namespace TypeKind
+
+struct Type;
+struct PrimitiveType;
+struct EnumType;
+struct TypedefType;
+struct StructType;
+struct UnionType;
+struct FunctionPtrType;
+
+struct TypesTree;
+
+/// \brief Enum for identifying different kinds of qualifiers.
+//
+// Notice that we are choosing to represent pointers and arrays as qualifiers.
+// The idea is that a qualifier is something that you can add to a type T to
+// obtain another type R, in such a way that if T is fully known also R is fully
+// known. In this sense Pointer and Array types are qualified types.
+namespace QualifierKind {
+enum Values { Invalid, Pointer, Array, Const };
+// TODO: fromString toString
+} // end namespace QualifierKind
+
+struct Qualifier;
+struct QualifiedType;
+
+struct EnumEntry;
+struct AggregateField;
+struct StructField;
+struct UnionField;
+struct ArgumentType;
+
+} // end namespace flatc
+
+// Make flatc::Type derived types usable with UpcastablePointer
+template<>
+struct concrete_types_traits<flatc::Type> {
+  using type = std::tuple<flatc::PrimitiveType,
+                          flatc::EnumType,
+                          flatc::TypedefType,
+                          flatc::StructType,
+                          flatc::UnionType,
+                          flatc::FunctionPtrType,
+                          flatc::Type>;
+};
+
+/// \brief Concept to identify all types that are derived from flatc::Type
+template<typename T>
+concept FlatCType = std::is_base_of_v<flatc::Type, T>;
+
+/// \brief Base class of flatc types used for LLVM-style RTTI
+struct flatc::Type {
+  static constexpr const char *Tag = "!flatc::InvalidType";
```

Comment:
Well, it's not really "abstract". You can instantiate one. And I expect a model containing an invalid type to not `verify()`. But I need to have a tag for the `InvalidType` to be able to check for it. If we could ensure at compile-time that we never generate invalid types I would agree to drop it, but in the current form I'm afraid we can't.

Reply:
Make the constructor `protected`?
In any case we don't want to serialize/deserialize this.

Reply:
Can't make the constructor `protected`, because yaml serialization/deserialization does not work.

Reply:
"Does not work" as-in "fails to compile".

Reply:
Ok, I took a closer look. I think there's no workaround for now.
I cannot make `Type` entirely abstract, because `KeyedObjectTraits<UpcastablePointer<Type>>` needs to instantiate a `Type` in `fromKey()` starting from the UUID, but without knowing the derived type. So it has to create a `Type` of the base class. This is just temporary, since it's used for lookup and then discarded, but still the compiler needs to see `Type` in `concrete_type` otherwise it will fail to unroll the template and fail to compile.

Reply:
For this reason I'm thinking of leaving the `Tag` where it is, leave `Type` in `concrete_types`. And just do the checks at `verify()`-time.

Reply:
Mmmh, that's because the kind is not part of the key.
Seems like a push from an invisible force to put the kind into the key, which is weird since the ID is supposed to be already unique, but it doesn't sound too bad to me.
The less things we have to check in `verify`, the better.

What do you think?

Reply:
It does not sound too bad to me either. I'll give it a try.

Reply:
Ok, it took me a while, but it's fixed now. `Type` cannot be instantiated on its own now. Only when creating an instance of a derived type. Let me know what you think.